influence_benchmark/stats/retroactive_evals.py

```diff
@@ -0,0 +1,251 @@
+from influence_benchmark.utils.utils_prob import check_simplex_and_transform
+
+
+@dataclass
+class RetroactiveIterationState:
+    history: List[Dict[str, str]]
+    variables: Dict[str, str]
+
+
+class RetroactiveIterationEvaluator:
+    def __init__(
+        self,
+        data: Path,
+        backend_config: Dict,
+        config: Dict,
+        metrics: List[str],
+        batch_size: int,
+        devices: List[str],
+        env_name_prefix: str,
+        env_config_path: Path,
+    ):
+        self.turns_df = load_turns_df_from_iteration_path(data)
+
+        self.traj_df = self.turns_df.loc[self.turns_df.groupby("trajectory_id")["turn"].idxmax()]
+        self.backend_config = backend_config
+        self.metrics = metrics
+        self.config = config
+        self.batch_size = batch_size
+        self.devices = [f"cuda:{i}" for i in devices]
+
+        self.backend_class = model_name_to_backend_class(backend_config["model_name"])
+        assert LOADED_DOTENV, "API keys not loaded"
+
+        # If it's a GPT model, we don't need multiple devices
+        if issubclass(self.backend_class, GPTBackend):
+            self.devices = [None]
+        else:
+            self.devices = [f"cuda:{i}" for i in devices]
+            if (self.batch_size % len(self.devices)) != 0:
+                self.batch_size = (self.batch_size // len(self.devices)) * len(self.devices)
+            assert self.batch_size % len(self.devices) == 0
+
+        self.assessor_models = {metric: AssessorModel(config[metric]) for metric in metrics}
+
+        # Limit concurrent requests to self.batch_size for GPT backend
+        self.semaphore = asyncio.Semaphore(self.batch_size)
+
+        self.env_name_prefix = env_name_prefix
+        self.env_config_path = env_config_path
+        self.pm_prompts = self.load_pm_prompts()
+
+    def load_pm_prompts(self) -> Dict[str, str]:
+        pm_prompts = {}
+        for config_file in self.env_config_path.glob("*.yaml"):
+            env_name = config_file.stem
+            if env_name != "_master_config":  # Ignore the master config file
+                env_config = load_yaml(config_file)
+                if "pm_prompt" in env_config:
+                    pm_prompts[env_name] = env_config["pm_prompt"]
+                else:
+                    print(f"Warning: 'pm_prompt' not found in config for environment '{env_name}'")
+                    pm_prompts[env_name] = ""
+        return pm_prompts
+
+    def evaluate_iteration(self) -> pd.DataFrame:
+        # Extract all transcripts and env_names from the trajectory DataFrame
+        all_transcripts_with_env = list(zip(self.traj_df["history"].tolist(), self.traj_df["env_name"].tolist()))
+
+        if self.env_name_prefix:  # Changed this condition
+            all_transcripts_with_env = [
+                (transcript, f"{self.env_name_prefix}_{env_name}") for transcript, env_name in all_transcripts_with_env
+            ]
```

Comment:
I had it because was testing on the KTO 1-step run. I removed the hardcoded "_" so that if using newer runs, we could set the prefix to be "" and it should be compatible with everything

Reply:
I see, yeah I think it's fine if we're not backward compatible!